File: cookiesync/cookie/domains.py
"""Host parsing and the cookie send-rule: what the browser would send to a host.

No public-suffix list: ``cookie_applies`` implements the actual domain-match the
browser uses, which is all we need to pick the cookies for one target host.
"""

from __future__ import annotations

from cookiesync.cookie.models import Host, HostKey


def normalize_host(url: str) -> Host:
    """Lowercase bare host from a URL or domain (strip scheme, path, query, port, leading dot)."""
    v = url.strip().lower()
    if "://" in v:
        v = v.split("://", 1)[1]
    v = v.split("/", 1)[0].split("?", 1)[0]
    if "@" in v:
        v = v.split("@", 1)[1]
    if ":" in v:
        v = v.split(":", 1)[0]
    return Host(v.strip("."))


def url_scheme(url: str, *, default: str = "https") -> str:
    """Scheme of a URL, or ``default`` for a bare domain."""
    return url.split("://", 1)[0].lower() if "://" in url else default


def cookie_applies(host_key: HostKey, host: Host) -> bool:
    """Would a browser send a cookie with this ``host_key`` to ``host``?

    Domain cookies (leading dot) match the base host and any subdomain; host-only
    cookies match exactly. Mirrors the browser's own send rule.
    """
    hk = host_key.lower()
    rh = host.lower()
    return (rh == hk[1:] or rh.endswith(hk)) if hk.startswith(".") else rh == hk

File: internal/cookie/domains.go
package cookie

import "strings"

// Host parsing and the cookie send-rule: what a browser would send to a host.
//
// There is no public-suffix list here: Applies implements the actual
// domain-match the browser uses, which is all we need to pick the cookies for one
// target host.

// NormalizeHost returns the bare, lowercase host of a URL or domain, stripping the
// scheme, path, query, port, userinfo, and any leading dot.
func NormalizeHost(url string) Host {
	v := strings.ToLower(strings.TrimSpace(url))
	if i := strings.Index(v, "://"); i >= 0 {
		v = v[i+len("://"):]
	}
	v, _, _ = strings.Cut(v, "/")
	v, _, _ = strings.Cut(v, "?")
	if _, after, found := strings.Cut(v, "@"); found {
		v = after
	}
	v, _, _ = strings.Cut(v, ":")
	return Host(strings.Trim(v, "."))
}

// URLScheme returns the lowercased scheme of a URL, or "https" for a bare domain.
func URLScheme(url string) string {
	if scheme, _, found := strings.Cut(url, "://"); found {
		return strings.ToLower(scheme)
	}
	return "https"
}

// Applies reports whether a browser would send a cookie with this hostKey to
// host. Domain cookies (leading dot) match the base host and any subdomain;
// host-only cookies match exactly. The comparison is case-insensitive.
func Applies(hostKey HostKey, host Host) bool {
	hk := strings.ToLower(string(hostKey))
	rh := strings.ToLower(string(host))
	if strings.HasPrefix(hk, ".") {
		return rh == hk[1:] || strings.HasSuffix(rh, hk)
	}
	return rh == hk
}
